Macros/Macros.js

```diff
@@ -33,54 +35,163 @@ class FurnaceMacros {
+                return sendResponse("Invalid macro type");
+            if (!Macros.canUseScripts(user))
+                return sendResponse(`You are not allowed to use JavaScript macros.`);
+            if (!macro.getFlag("furnace", "runAsGM") || !macro.canRunAsGM)
+                return sendResponse(`You are not authorized to run this macro as the GM.`);
+
+            const context = FurnaceMacros.getTemplateContext(message.args, message.context);
+            try {
+                const result = macro.callScriptFunction(context);
+                return sendResponse(null, result);
+            } catch (err) {
+                console.error(err);
+                return sendResponse(`There was an error in your macro syntax. See the console (F12) of the GM '${game.user.name}' for details`);
+            }
+        } else if (message.action === "GMMacroResult") {
+            const resolve = this._requestResolvers[message.requestId];
+            if (resolve) {
+                delete this._requestResolvers[message.requestId];
+                resolve({ result: message.result, error: message.error });
+            }
+        }
     }
 
-    static getTemplateContext(args=null) {
-        if (args && args[0] === "RemoteContext") {
-          const token = canvas.tokens.get(args[1].tokenId)
-          const actor = token.actor;
-          const sceneId = args[1].sceneId || canvas.scene.id; 
-        const speaker = { scene: sceneId, actor: actor.id, token: args[1].tokenId, alias: token.actor.name}
-          const event = args[1].event;
-          return { game: game, ui: ui, canvas: canvas, speaker, actor, token, character: actor, sceneId, event, args: args.slice(2)}
+    static getTemplateContext(args = null, remoteContext = null) {
+        const context = {
+            game: game,
+            ui: ui,
+            canvas: canvas,
+            scene: canvas.scene,
+            args,
+            speaker: {},
+            actor: null,
+            token: null,
+            character: null
+        };
+        if (remoteContext) {
+            // Set the context based on the remote context, and make sure data is valid and the remote
+            // has a token/actor selected.
+            context.speaker = remoteContext.speaker || {};
+            if (remoteContext.actorId)
+                context.actor = game.actors.get(remoteContext.actorId) || null;
+            if (remoteContext.sceneId)
+                context.scene = game.scenes.get(remoteContext.sceneId) || canvas.scene;
+            if (remoteContext.tokenId) {
+                if (canvas.scene.id === context.scene.id) {
+                    context.token = canvas.tokens.get(remoteContext.tokenId) || null;
+                } else {
+                    const tokenData = context.scene.getEmbeddedEntity("Token", remoteContext.tokenId);
+                    if (tokenData)
+                        context.token = new Token(tokenData, context.scene);
```

Comment:
for the created token, token.x and token.y are zero. But token.center is fine so I din't think it is worth worrying about

Reply:
That didn't make sense until I checked the code, there is no 'x' and 'y' getters, and a Token is a `PlaceableObject` (so a `PIXI.Container`), which means the x and y are actually PIXI object positions.
As long as the token is not added to the scene and drawn, they won't be set I guess. The `.center` is a getter which uses `this.data.x` and `this.data.y` which is why it works.
I guess this means we really should always use `token.data.x/y` rather than relying on the `token.x/y`
Good catch!